packages/react/src/index.ts

```diff
@@ -15,64 +17,165 @@ import {
 
-function setCurrentUpdater(updater?: Effect) {
-	// end tracking for the current update:
-	if (finishUpdate) finishUpdate();
-	// start tracking the new update:
-	finishUpdate = updater && updater._start();
+		return ProxyComponent;
+	}
+
+	/**
+	 * Emulate a Proxy if environment doesn't support it.
+	 *
+	 * @TODO - unlike Proxy, it's not possible to access the type/Component's
+	 * static properties this way. Not sure if we want to copy all statics here.
+	 * Omitting this for now.
+	 *
+	 * @example - works with Proxy, doesn't with wrapped function.
+	 * ```
+	 * const el = <SomeFunctionalComponent />
+	 * el.type.someOwnOrInheritedProperty;
+	 * el.type.defaultProps;
+	 * ```
+	 */
+	const WrappedComponent = function () {
+		return ProxyHandlers.apply(Component, undefined, arguments);
+	};
+	ProxyInstance.set(Component, WrappedComponent);
+	ProxyInstance.set(WrappedComponent, WrappedComponent);
+
+	return WrappedComponent;
 }
 
-function createUpdater(update: () => void) {
+/**
+ * A redux-like store whose store value is a positive 32bit integer (a 'version').
+ *
+ * React subscribes to this store and gets a snapshot of the current 'version',
+ * whenever the 'version' changes, we tell React it's time to update the component (call 'onStoreChange').
+ *
+ * How we achieve this is by creating a binding with an 'effect', when the `effect._callback' is called,
+ * we update our store version and tell React to re-render the component ([1] We don't really care when/how React does it).
+ *
+ * [1]
+ * @see https://reactjs.org/docs/hooks-reference.html#usesyncexternalstore
+ * @see https://github.com/reactjs/rfcs/blob/main/text/0214-use-sync-external-store.md
+ */
+function createEffectStore() {
 	let updater!: Effect;
-	effect(function (this: Effect) {
+	let version = 0;
+	let onChangeNotifyReact: (() => void) | undefined;
+
+	let unsubscribe = effect(function (this: Effect) {
 		updater = this;
 	});
-	updater._callback = update;
-	return updater;
+	updater._callback = function () {
+		if (!onChangeNotifyReact) return void unsubscribe();
+
+		version = (version + 1) | 0;
+		onChangeNotifyReact!();
+	};
+
+	return {
+		updater,
+		subscribe(onStoreChange: () => void) {
+			onChangeNotifyReact = onStoreChange;
+
+			return function () {
+				/**
+				 * @todo - we may want to unsubscribe here instead?
+				 * Components wrapped in `memo` no longer get updated if we unsubscribe here.
+				 */
+				onChangeNotifyReact = undefined;
+			};
+		},
+		getSnapshot() {
+			return version;
+		},
+	};
 }
 
+function WrapJsx<T>(jsx: T): T {
+	if (typeof jsx !== "function") return jsx;
+
+	return function (type: any, props: any, ...rest: any[]) {
+		if (typeof type === "function" && !(type instanceof Component)) {
+			return jsx.call(jsx, ProxyFunctionalComponent(type), props, ...rest);
+		}
+
+		if (type && typeof type === "object" && type.$$typeof === ReactMemoType) {
+			return jsx.call(jsx, ProxyFunctionalComponent(type.type), props, ...rest);
```

Comment:
This should set `type.type = ProxyFunctionalComponent(...)` and not unwrap the memo component 🤦